modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientRequest.java

```diff
@@ -58,4 +59,21 @@ public ClientRequest(long reqId) {
     public ClientResponse process(ClientConnectionContext ctx) {
         return new ClientResponse(reqId);
     }
+
+    /**
+     * Processes the request asynchronously.
+     *
+     * @return Future for response.
+     */
+    public IgniteInternalFuture<ClientResponse> processAsync(ClientConnectionContext ctx) {
```

Comment:
Have you considered just replacing the "process" method with "processAsync"?

For the first glance it might help avoid creating `isAsync` method and ClientListenerAsyncResponse. It looks a bit confusing that it's up to a outer code to decide whether to run `sync` or `async` method to process `ClientRequest` while `ClientRequest` itself knows how it should be processed.

On the other hand it will lead to creating CompletedFuture wrappers and a bit of refactoring.